Profiles

arthurzam opened the pull request:
The feature implemented in this pull request is changing profiles (settings group). For example, to have a profile for videos (like the placement of docks, equalizer) and another one for music files.

In the background it is implemented using sub directories in side `~/.qmplay/profiles/` (or the default one). There is full backward compatibility (the old settings is called default).

As always, please check the code style, tr function use (maybe I missed somewhere) and that it works on your systems.

zaps166 commented:
I thought about this feature, thanks! :smile:

Please add spaces after `if` and loop operators (I love spaces after them :) ).

A few questions:
- After creation of the new profile it is not visible on the list soon after creation. Probably minor bug.
- If I create a new profile it resets the layout to default settings. Shouldn't it copy the current layout instead? Maybe add three sub-menus: Copy layout, create a new layout and reset layout?
- How to remove or update the current profile?

arthurzam commented:
> Please add spaces after if and loop operators (I love spaces after them :) ).

I hope I fixed all of them.

> After creation of the new profile it is not visible on the list soon after creation. Probably minor bug.

Fixed 😄 

> If I create a new profile it resets the layout to default settings. Shouldn't it copy the current layout instead? Maybe add three sub-menus: Copy layout, create a new layout and reset layout?

I added an optional "copy profile" action. This copies all your current settings, enabling you to "fork" your current profile.
About the layout deal, this is part of the settings. Therefore, an reset layout action should be in another menu, not connected to the profiles submenu at all.

> How to remove or update the current profile?

For now not implemented. As the implementation is easy (if the profile is default - only show message box that it is impossible, otherwise just delete `~/.qmplay/profiles/${profileName}/` directory.
The problem how to add it to the User Interface. Your thoughts?

zaps166 commented:
> How to remove or update the current profile?

Two ways:
1. Right click on profile menu-item and then show popup with "Remove" menu-item (just like removing equalizer presets).
2. In Settings right of profiles ComboBox, add a "Remove" button.

> if the profile is default - only show message box that it is impossible

Or even don't show the popup menu and disable the button in settings :)

---

> Therefore, an reset layout action should be in another menu, not connected to the profiles submenu at all.

Ok, not necessary :smile:

---

`~/.qmplay/profiles/` - All other directories starts from capital letter, so this also should start from capital letter for consistency :)

---

The Profiles menu-item should have a separator after settings menu-item.

arthurzam commented:
> Two ways...
> Or even don't show the popup menu and disable the button in settings
> All other directories starts from capital letter, so this also should start from capital letter for consistency
> The Profiles menu-item should have a separator after settings menu-item.

Done 😃

zaps166 commented on src/gui/MenuBar.cpp:
```diff
@@ -466,6 +494,89 @@ void MenuBar::setKeyShortcuts()
+	{
+		const QString &profile = act->property("path").toString();
+		Functions::dirRemoveRecursively(QMPlay2Core.getSettingsDir() + "Profiles/" + profile);
+		if("Profiles/" + profile + "/" == QMPlay2Core.getSettingsProfile())
```
:smile:

zaps166 commented on src/gui/MenuBar.cpp:
```diff
@@ -466,6 +494,89 @@ void MenuBar::setKeyShortcuts()
+	if (selectedProfile != profileSettings.value("Profile", "/").toString())
+	{
+		QDir(QMPlay2Core.getSettingsDir()).mkpath("Profiles/" + selectedProfile);
+		QFile tempFile(QMPlay2Core.getSettingsDir() + "Profiles/" + selectedProfile + "/QMPlay2.ini");
```
Is it necessary?

zaps166 commented on src/gui/MenuBar.cpp:
```diff
@@ -466,6 +494,89 @@ void MenuBar::setKeyShortcuts()
+		QMPlay2GUI.mainW->close();
+	}
+}
+void MenuBar::copyProfile()
```
`copyProfile()` and `addProfile()` can share the same code e.g. in `static bool ...()` function.

zaps166 commented on src/gui/MenuBar.cpp:
```diff
@@ -466,6 +494,89 @@ void MenuBar::setKeyShortcuts()
+		{
+			QSettings profileSettings(QMPlay2Core.getSettingsDir() + "Profile.ini", QSettings::IniFormat);
+			profileSettings.setValue("Profile", "/");
+			QMPlay2GUI.restartApp = QMPlay2GUI.removeSettings = true;
```
`QMPlay2GUI.removeSettings` also removes settings, so is the `Functions::dirRemoveRecursively` necessary? maybe simple `rmdir` would be enough after profile removal?

zaps166 commented on src/qmplay2/Functions.cpp:
```diff
@@ -612,3 +633,19 @@ bool Functions::wrapMouse(QWidget *widget, QPoint &mousePos, int margin)
+{
+	QStringList profiles;
+
+	QDirIterator it(QMPlay2Core.getSettingsDir() + "Profiles/", QStringList() << "QMPlay2.ini", QDir::Files, QDirIterator::Subdirectories);
```
Filtering `QMPlay2.ini` shouldn't be necessary, just show only available directories. This will prevent creating an empty file. Empty directory should be a clean, new profile. QMPlay2 will automatically creates `ini` files :)

arthurzam commented:
> Default profile should start from capital letter also in settings widget in ComboBox
> I'm able to remove the default profile from ComboBox.
> Filtering QMPlay2.ini
> copyProfile() and addProfile() can share the same code e.g. in static bool ...() function.
> maybe simple rmdir would be enough after profile removal?
> ...

all fixed 😃 
please review

zaps166 commented on src/gui/SettingsWidget.cpp:
```diff
@@ -309,6 +310,21 @@ SettingsWidget::SettingsWidget(int page, const QString &moduleName, QWidget *vid
 				page1->subsLangB->setCurrentIndex(page1->subsLangB->count() - 1);
 		}
+		{
+			page1->profileB->addItem("Default", "/");
```
Missing `tr()`

zaps166 commented on src/gui/MenuBar.cpp:
```diff
@@ -466,6 +494,88 @@ void MenuBar::setKeyShortcuts()
+	if (QAction *act = (QAction *)options->removeProfileMenu->property("profile").value<void *>())
+	{
+		const QString &profile = "Profiles/" + act->property("path").toString() + "/";
+		if(profile == QMPlay2Core.getSettingsProfile())
```
space :D

zaps166 commented on src/gui/SettingsWidget.cpp:
```diff
@@ -954,3 +978,26 @@ void SettingsWidget::resetSettings()
+	if (selectedProfile != "/")
+		QDir(QMPlay2Core.getSettingsDir()).rmdir(selectedProfile);
+
+	if(selectedProfile == QMPlay2Core.getSettingsProfile())
```
space :D

arthurzam commented:
> 3 minor fixes needed

Done
